Controller/Adminhtml/Menu/Save.php

```diff
@@ -102,107 +102,86 @@ public function execute()
         $menu->saveStores($this->getRequest()->getParam('stores'));
         $nodes = $this->getRequest()->getParam('serialized_nodes');
 
-        if (!empty($nodes)) {
-            $nodes = json_decode($nodes, true);
-            $nodes = $this->_convertTree($nodes, '#');
-
-            if (!empty($nodes)) {
-                $filterBuilder = $this->filterBuilderFactory->create();
-                $filter = $filterBuilder->setField('menu_id')
-                    ->setValue($menu->getMenuId())
-                    ->setConditionType('eq')
-                    ->create();
-
-                $filterGroupBuilder = $this->filterGroupBuilderFactory->create();
-                $filterGroup = $filterGroupBuilder->addFilter($filter)->create();
-
-                $searchCriteriaBuilder = $this->searchCriteriaBuilderFactory->create();
-                $searchCriteria = $searchCriteriaBuilder->setFilterGroups([$filterGroup])->create();
-
-                $oldNodes = $this->nodeRepository->getList($searchCriteria)->getItems();
-
-                $existingNodes = [];
-
-                foreach ($oldNodes as $node) {
-                    $existingNodes[$node->getId()] = $node;
-                }
-
-                $nodesToDelete = [];
-
-                foreach ($existingNodes as $nodeId => $noe) {
-                    $nodesToDelete[$nodeId] = true;
-                }
-
-                $nodeMap = [];
+        $existingNodes = [];
+        foreach ($this->getCurrentNodes($menu) as $node) {
+            $existingNodes[$node->getId()] = $node;
+        }
 
-                foreach ($nodes as $node) {
-                    $nodeId = $node['id'];
+        $nodesToDelete = [];
+        foreach ($existingNodes as $nodeId => $node) {
+            $nodesToDelete[$nodeId] = true;
+        }
 
-                    if (in_array($nodeId, array_keys($existingNodes))) {
-                        unset($nodesToDelete[$nodeId]);
-                        $nodeMap[$nodeId] = $existingNodes[$nodeId];
-                    } else {
-                        $nodeObject = $this->nodeFactory->create();
-                        $nodeObject->setMenuId($menu->getMenuId());
-                        $nodeObject = $this->nodeRepository->save($nodeObject);
-                        $nodeMap[$nodeId] = $nodeObject;
-                    }
-                }
+        $nodes = $nodes ? json_decode($nodes, true) : [];
+        $nodes = $this->_convertTree($nodes, '#');
+        $nodeMap = [];
+
+        foreach ($nodes as $node) {
+            $nodeId = $node['id'];
+
+            if (in_array($nodeId, array_keys($existingNodes))) {
```

Comment:
`if(isset($existingNodes[$nodeId]))` seems to be more performant way to do this check